src/Network/NWBrowser.cs

```diff
@@ -113,22 +121,65 @@ public NWParameters Parameters
 		[DllImport (Constants.NetworkLibrary)]
 		unsafe static extern void nw_browser_set_browse_results_changed_handler (OS_nw_browser browser, void *handler);
 
-		delegate void nw_browser_browse_results_changed_handler_t (IntPtr block, IntPtr oldResult, IntPtr newResult);
+		delegate void nw_browser_browse_results_changed_handler_t (IntPtr block, IntPtr oldResult, IntPtr newResult, bool completed);
 		static nw_browser_browse_results_changed_handler_t static_ChangesHandler = TrampolineChangesHandler;
 
 		[MonoPInvokeCallback (typeof (nw_browser_browse_results_changed_handler_t))]
-		static void TrampolineChangesHandler (IntPtr block, IntPtr oldResult, IntPtr newResult)
+		static void TrampolineChangesHandler (IntPtr block, IntPtr oldResult, IntPtr newResult, bool completed)
 		{
-			var del = BlockLiteral.GetTarget<Action<NWBrowseResult, NWBrowseResult>> (block);
+			var del = BlockLiteral.GetTarget<NWBrowserChangesDelegate> (block);
 			if (del != null) {
-				using (var nwOldResult = new NWBrowseResult (oldResult, owns: false))
-				using (var nwNewResult = new NWBrowseResult (newResult, owns: false))
-					del (nwOldResult, nwNewResult);
+				// we do the cleanup of the objs in the internal handlers
+				NWBrowseResult nwOldResult = (oldResult == IntPtr.Zero)? null : new NWBrowseResult (oldResult, owns: false);
+				NWBrowseResult nwNewResult = (newResult == IntPtr.Zero)? null : new NWBrowseResult (newResult, owns: false);
+				del (nwOldResult, nwNewResult, completed);
+			}
+		}
+
+		public Action<NWBrowseResult, NWBrowseResult> IndividualChangesDelegate { get; set; }
+
+		// syntactic sugar for the user, nicer to get all the changes at once
+		public NWBrowserCompleteChangesDelegate CompleteChangesDelegate { get; set; }
+		object changesLock = new object ();
+		List<(NWBrowseResult result, NWBrowseResultChange change)> changes = new List<(NWBrowseResult result, NWBrowseResultChange change)> ();
+
+		void InternalChangesHandler (NWBrowseResult oldResult, NWBrowseResult newResult, bool completed)
+		{
+			// we allow the user to listen to both, individual changes AND complete ones, individual is simple, just
+			// call the cb, completed, we need to get a collection and call the cb when completed
+			var individualCb = IndividualChangesDelegate;
+			individualCb?.Invoke (oldResult, newResult);
+			lock (changesLock) {
```

Comment:
The lock is surrounding a rather big block of code, if I understand correctly you only want to lock around accesses to the `changes` variable.

My suggestion would be to only lock when accessing the `changes` variable, and also capture the variable and hoist it outside of the lock statement when invoking the complete handler and disposing the changes (this would start where you're currently calling `changes.Add (result)`):

```csharp

List<(NWBrowseResult result, NWBrowseResultChange change)> tmp_changes;
lock (changesLock) {
    changes.Add (result);
    if (complete) {
        tmp_changes = changes;
        changes = new List<(NWBrowseResult result, NWBrowseResultChange change)> ();
    }
}
if (completed) {
    completeCb.Invoke (tmp_changes);
    foreach (var c in tmp_changes)
        c.result?.Dispose ();
}
```

In general the code inside locks should be as short as possible, and we should really try to avoid executing code we can't control (such as calling user delegates).